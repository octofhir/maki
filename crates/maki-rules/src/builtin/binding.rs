//! Binding strength validation rules
//!
//! Validates that bindings to value sets have proper strength specifications
//! and use valid strength values.

use maki_core::cst::ast::{AstNode, Document, ValueSetRule};
use maki_core::{Diagnostic, SemanticModel, Severity};

/// Rule ID for binding strength validation
pub const BINDING_STRENGTH_PRESENT: &str = "blocking/binding-strength-present";

/// Valid binding strengths in FHIR
const VALID_BINDING_STRENGTHS: &[&str] = &["required", "extensible", "preferred", "example"];

/// Check for missing or invalid binding strengths in FSH document
pub fn check_binding_strength(model: &SemanticModel) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();

    let Some(document) = Document::cast(model.cst.clone()) else {
        return diagnostics;
    };

    // Check profiles
    for profile in document.profiles() {
        diagnostics.extend(check_value_set_bindings(
            profile.rules().filter_map(|r| match r {
                maki_core::cst::ast::Rule::ValueSet(vs) => Some(vs),
                _ => None,
            }),
            model,
        ));
    }

    // Check extensions
    for extension in document.extensions() {
        diagnostics.extend(check_value_set_bindings(
            extension.rules().filter_map(|r| match r {
                maki_core::cst::ast::Rule::ValueSet(vs) => Some(vs),
                _ => None,
            }),
            model,
        ));
    }

    diagnostics
}

/// Check value set binding rules for proper strength specification
fn check_value_set_bindings(
    rules: impl Iterator<Item = ValueSetRule>,
    model: &SemanticModel,
) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();

    for rule in rules {
        match rule.strength() {
            Some(strength) => {
                // Check if the strength is valid
                let strength_lower = strength.to_lowercase();
                if !VALID_BINDING_STRENGTHS.contains(&strength_lower.as_str()) {
                    let location = model.source_map.node_to_diagnostic_location(
                        rule.syntax(),
                        &model.source,
                        &model.source_file,
                    );
                    diagnostics.push(
                        Diagnostic::new(
                            BINDING_STRENGTH_PRESENT,
                            Severity::Error,
                            format!(
                                "Invalid binding strength '{strength}'. Must be one of: required, extensible, preferred, example"
                            ),
                            location,
                        )
                        .with_code("invalid-binding-strength".to_string()),
                    );
                }
            }
            None => {
                // Missing binding strength
                if let Some(value_set_name) = rule.value_set() {
                    let location = model.source_map.node_to_diagnostic_location(
                        rule.syntax(),
                        &model.source,
                        &model.source_file,
                    );
                    diagnostics.push(
                        Diagnostic::new(
                            BINDING_STRENGTH_PRESENT,
                            Severity::Error,
                            format!(
                                "Binding to ValueSet '{value_set_name}' must specify a strength (required, extensible, preferred, or example)"
                            ),
                            location,
                        )
                        .with_code("missing-binding-strength".to_string()),
                    );
                }
            }
        }
    }

    diagnostics
}
