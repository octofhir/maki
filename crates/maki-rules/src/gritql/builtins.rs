//! FSH-specific GritQL built-in functions - CST-based
//!
//! This module provides custom built-in functions for GritQL patterns that are specific to FSH.
//! These functions can be used in .grit pattern files to perform FSH-aware queries and validations.

use super::cst_adapter::FshGritNode;
use maki_core::cst::FshSyntaxKind;

/// Register all FSH-specific built-in functions
///
/// These functions extend GritQL with FSH domain knowledge:
/// - `is_profile(node)` - Check if a node is a Profile definition
/// - `is_extension(node)` - Check if a node is an Extension definition
/// - `is_value_set(node)` - Check if a node is a ValueSet definition
/// - `is_code_system(node)` - Check if a node is a CodeSystem definition
/// - `has_comment(node)` - Check if a node has a comment
pub fn register_fsh_builtins() -> Vec<&'static str> {
    vec![
        "is_profile",
        "is_extension",
        "is_value_set",
        "is_code_system",
        "has_comment",
    ]
}

/// Check if a node is a Profile definition
pub fn is_profile(node: &FshGritNode) -> bool {
    node.kind() == FshSyntaxKind::Profile
}

/// Check if a node is an Extension definition
pub fn is_extension(node: &FshGritNode) -> bool {
    node.kind() == FshSyntaxKind::Extension
}

/// Check if a node is a ValueSet definition
pub fn is_value_set(node: &FshGritNode) -> bool {
    node.kind() == FshSyntaxKind::ValueSet
}

/// Check if a node is a CodeSystem definition
pub fn is_code_system(node: &FshGritNode) -> bool {
    node.kind() == FshSyntaxKind::CodeSystem
}

/// Check if a node or its children contain comments
pub fn has_comment(node: &FshGritNode) -> bool {
    use grit_util::AstNode;

    // Check if current node is a comment
    if matches!(
        node.kind(),
        FshSyntaxKind::CommentLine | FshSyntaxKind::CommentBlock
    ) {
        return true;
    }

    // Check children for comments
    node.children().any(|child| has_comment(&child))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::gritql::FshGritTree;
    use grit_util::{Ast, AstNode};

    #[test]
    fn test_register_builtins() {
        let builtins = register_fsh_builtins();
        assert!(builtins.contains(&"is_profile"));
        assert!(builtins.contains(&"is_extension"));
        assert!(builtins.contains(&"is_value_set"));
        assert!(builtins.contains(&"has_comment"));
    }

    #[test]
    fn test_is_profile() {
        let source = "Profile: MyPatient\nParent: Patient";
        let tree = FshGritTree::parse(source);
        let root = tree.root_node();

        // Find profile node in children
        let profile = root.children().find(|n| n.kind() == FshSyntaxKind::Profile);

        if let Some(profile_node) = profile {
            assert!(is_profile(&profile_node));
        }
    }

    #[test]
    fn test_is_value_set() {
        let source = "ValueSet: MyVS";
        let tree = FshGritTree::parse(source);
        let root = tree.root_node();

        let value_set = root
            .children()
            .find(|n| n.kind() == FshSyntaxKind::ValueSet);

        if let Some(vs_node) = value_set {
            assert!(is_value_set(&vs_node));
            assert!(!is_profile(&vs_node));
        }
    }
}
