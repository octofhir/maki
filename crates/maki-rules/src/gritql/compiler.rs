//! GritQL pattern compiler
//!
//! Compiles parsed GritQL AST into executable grit-pattern-matcher Pattern structs.

use super::parser::{GritPattern, GritPredicate};
use super::query_context::{FshNodePattern, FshQueryContext};
use grit_pattern_matcher::pattern::{
    And, Container, Not, Pattern, PrAnd, PrNot, PrOr, Predicate, Variable, Where,
};
use maki_core::cst::FshSyntaxKind;
use maki_core::{MakiError, Result};
use std::collections::HashMap;

pub struct PatternCompiler {
    variables: HashMap<String, usize>,
    next_var_index: usize,
    current_scope: usize,
}

impl PatternCompiler {
    pub fn new() -> Self {
        Self {
            variables: HashMap::new(),
            next_var_index: 0,
            current_scope: 0,
        }
    }

    pub fn compile(&mut self, pattern: &GritPattern) -> Result<Pattern<FshQueryContext>> {
        match pattern {
            GritPattern::NodeKind(kind) => self.compile_node_kind(kind),
            GritPattern::Variable(name) => self.compile_variable(name),
            GritPattern::Not(inner) => self.compile_not(inner),
            GritPattern::Where(pattern, predicate) => self.compile_where(pattern, predicate),
            GritPattern::Assignment { var, value } => self.compile_assignment(var, value),
            GritPattern::And(patterns) => self.compile_and(patterns),
            GritPattern::Or(patterns) => self.compile_or(patterns),
        }
    }

    fn compile_node_kind(&self, kind: &str) -> Result<Pattern<FshQueryContext>> {
        let syntax_kind = self.parse_syntax_kind(kind)?;
        Ok(Pattern::AstNode(Box::new(FshNodePattern {
            kind: syntax_kind,
            args: vec![],
        })))
    }

    fn compile_variable(&mut self, name: &str) -> Result<Pattern<FshQueryContext>> {
        let index = self.get_or_create_variable(name);
        Ok(Pattern::Variable(Variable::new(index, self.current_scope)))
    }

    fn compile_not(&mut self, inner: &GritPattern) -> Result<Pattern<FshQueryContext>> {
        let compiled_inner = self.compile(inner)?;
        Ok(Pattern::Not(Box::new(Not {
            pattern: compiled_inner,
        })))
    }

    fn compile_where(
        &mut self,
        pattern: &GritPattern,
        predicate: &GritPredicate,
    ) -> Result<Pattern<FshQueryContext>> {
        let compiled_pattern = self.compile(pattern)?;
        let compiled_predicate = self.compile_predicate_to_predicate(predicate)?;

        Ok(Pattern::Where(Box::new(Where {
            pattern: compiled_pattern,
            side_condition: compiled_predicate,
        })))
    }

    fn compile_assignment(
        &mut self,
        var: &str,
        value: &GritPattern,
    ) -> Result<Pattern<FshQueryContext>> {
        let var_index = self.get_or_create_variable(var);
        let compiled_value = self.compile(value)?;

        Ok(Pattern::Assignment(Box::new(
            grit_pattern_matcher::pattern::Assignment {
                container: Container::Variable(Variable::new(var_index, self.current_scope)),
                pattern: compiled_value,
            },
        )))
    }

    fn compile_and(&mut self, patterns: &[GritPattern]) -> Result<Pattern<FshQueryContext>> {
        let compiled: Result<Vec<_>> = patterns.iter().map(|p| self.compile(p)).collect();
        Ok(Pattern::And(Box::new(And {
            patterns: compiled?,
        })))
    }

    fn compile_or(&mut self, patterns: &[GritPattern]) -> Result<Pattern<FshQueryContext>> {
        let compiled: Result<Vec<_>> = patterns.iter().map(|p| self.compile(p)).collect();
        Ok(Pattern::Or(Box::new(grit_pattern_matcher::pattern::Or {
            patterns: compiled?,
        })))
    }

    #[allow(clippy::only_used_in_recursion)]
    fn compile_predicate_to_predicate(
        &mut self,
        predicate: &GritPredicate,
    ) -> Result<Predicate<FshQueryContext>> {
        match predicate {
            GritPredicate::FieldExists(_field) => Ok(Predicate::True),
            GritPredicate::Not(inner) => {
                let compiled = self.compile_predicate_to_predicate(inner)?;
                Ok(Predicate::Not(Box::new(PrNot::new(compiled))))
            }
            GritPredicate::And(preds) => {
                let compiled: Result<Vec<_>> = preds
                    .iter()
                    .map(|p| self.compile_predicate_to_predicate(p))
                    .collect();
                Ok(Predicate::And(Box::new(PrAnd::new(compiled?))))
            }
            GritPredicate::Or(preds) => {
                let compiled: Result<Vec<_>> = preds
                    .iter()
                    .map(|p| self.compile_predicate_to_predicate(p))
                    .collect();
                Ok(Predicate::Or(Box::new(PrOr::new(compiled?))))
            }
        }
    }

    fn parse_syntax_kind(&self, kind: &str) -> Result<FshSyntaxKind> {
        let normalized = kind.to_uppercase().replace('-', "_");

        match normalized.as_str() {
            "PROFILE" => Ok(FshSyntaxKind::Profile),
            "EXTENSION" => Ok(FshSyntaxKind::Extension),
            "VALUESET" | "VALUE_SET" => Ok(FshSyntaxKind::ValueSet),
            "CODESYSTEM" | "CODE_SYSTEM" => Ok(FshSyntaxKind::CodeSystem),
            "INSTANCE" => Ok(FshSyntaxKind::Instance),
            "INVARIANT" => Ok(FshSyntaxKind::Invariant),
            "MAPPING" => Ok(FshSyntaxKind::Mapping),
            "LOGICAL" => Ok(FshSyntaxKind::Logical),
            "RESOURCE" => Ok(FshSyntaxKind::Resource),
            "ALIAS" => Ok(FshSyntaxKind::Alias),
            "PARENT" => Ok(FshSyntaxKind::ParentKw),
            "ID" => Ok(FshSyntaxKind::IdKw),
            "TITLE" => Ok(FshSyntaxKind::TitleKw),
            "DESCRIPTION" => Ok(FshSyntaxKind::DescriptionKw),
            _ => Err(MakiError::rule_error(
                "gritql-compiler",
                format!("Unknown FSH syntax kind: {kind}"),
            )),
        }
    }

    fn get_or_create_variable(&mut self, name: &str) -> usize {
        if let Some(&index) = self.variables.get(name) {
            index
        } else {
            let index = self.next_var_index;
            self.variables.insert(name.to_string(), index);
            self.next_var_index += 1;
            index
        }
    }
}

impl Default for PatternCompiler {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::gritql::parser::GritQLParser;

    #[test]
    fn test_compile_simple_node() {
        let mut parser = GritQLParser::new("profile");
        let pattern = parser.parse().unwrap();

        let mut compiler = PatternCompiler::new();
        let result = compiler.compile(&pattern);
        assert!(result.is_ok());
    }

    #[test]
    fn test_compile_where_clause() {
        let mut parser = GritQLParser::new("profile where { description }");
        let pattern = parser.parse().unwrap();

        let mut compiler = PatternCompiler::new();
        let result = compiler.compile(&pattern);
        assert!(result.is_ok());
    }

    #[test]
    fn test_compile_variable() {
        let mut parser = GritQLParser::new("$name");
        let pattern = parser.parse().unwrap();

        let mut compiler = PatternCompiler::new();
        let result = compiler.compile(&pattern);
        assert!(result.is_ok());
    }
}
