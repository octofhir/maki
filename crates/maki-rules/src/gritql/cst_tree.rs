//! FSH GritQL Tree - CST-based implementation
//!
//! This module provides the `Ast` trait implementation that allows GritQL
//! to query our Rowan-based CST. Unlike the old implementation that converted
//! a flat AST to a tree, this one directly exposes the CST which already
//! has the tree structure we need.

use super::cst_adapter::FshGritNode;
use grit_util::Ast;
use maki_core::cst::{FshSyntaxNode, parse_fsh};
use std::borrow::Cow;

/// FSH GritQL Tree - implements grit_util::Ast over our Rowan CST
///
/// This is a lightweight wrapper that just holds the root node and source text.
/// The actual tree structure is provided by Rowan.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FshGritTree {
    /// Root CST node wrapped for GritQL
    root: FshGritNode,
    /// Source text (owned for lifetime management)
    source: String,
}

impl FshGritTree {
    /// Parse FSH source and create a GritQL-queryable tree
    ///
    /// # Example
    ///
    /// ```ignore
    /// use maki_rules::gritql::FshGritTree;
    ///
    /// let source = "Profile: MyPatient\nParent: Patient";
    /// let tree = FshGritTree::parse(source);
    ///
    /// // Now you can query this tree with GritQL patterns
    /// ```
    pub fn parse(source: &str) -> Self {
        let (cst, _errors) = parse_fsh(source);
        Self::from_cst(cst, source.to_string())
    }

    /// Create a GritQL tree from an existing CST node
    ///
    /// This is useful when you've already parsed the FSH and want to
    /// run GritQL queries on a specific subtree.
    pub fn from_cst(root: FshSyntaxNode, source: String) -> Self {
        let root = FshGritNode::new(root);
        Self { root, source }
    }

    /// Alias for from_cst (for backwards compatibility with tests)
    pub fn new(root: FshSyntaxNode, source: String) -> Self {
        Self::from_cst(root, source)
    }

    /// Get the root node
    pub fn root(&self) -> &FshGritNode {
        &self.root
    }

    /// Get the source text
    pub fn source_text(&self) -> &str {
        &self.source
    }
}

impl Ast for FshGritTree {
    type Node<'a> = FshGritNode;

    fn root_node(&self) -> Self::Node<'_> {
        // Cheap clone (Rowan uses Arc)
        self.root.clone()
    }

    fn source(&self) -> Cow<'_, str> {
        Cow::Borrowed(&self.source)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use grit_util::{Ast, AstCursor, AstNode};

    #[test]
    fn test_parse_simple_profile() {
        let source = "Profile: MyPatient\nParent: Patient";
        let tree = FshGritTree::parse(source);

        let root = tree.root_node();
        assert!(root.children().count() > 0, "Root should have children");
    }

    #[test]
    fn test_source_preservation() {
        let source = "Profile: MyPatient  // Comment\nParent: Patient";
        let tree = FshGritTree::parse(source);

        // Lossless property: tree should contain exact source
        assert_eq!(tree.source(), source);
    }

    #[test]
    fn test_ast_trait_root_node() {
        let source = "Profile: Test";
        let tree = FshGritTree::parse(source);

        let root = tree.root_node();
        let text = root.text().unwrap();

        assert_eq!(text.as_ref(), source);
    }

    #[test]
    fn test_from_cst() {
        let source = "Profile: MyPatient";
        let (cst, _) = parse_fsh(source);

        let tree = FshGritTree::from_cst(cst, source.to_string());

        assert_eq!(tree.source(), source);
        assert!(tree.root().children().count() > 0);
    }

    #[test]
    fn test_multiple_definitions() {
        let source = r#"Profile: Patient1
Parent: Patient

Profile: Patient2
Parent: Patient

ValueSet: TestVS
"#;
        let tree = FshGritTree::parse(source);

        let root = tree.root_node();
        let children: Vec<_> = root.children().collect();

        assert!(
            children.len() >= 3,
            "Should have multiple top-level definitions"
        );
    }

    #[test]
    fn test_tree_navigation() {
        let source = "Profile: MyPatient\nParent: Patient";
        let tree = FshGritTree::parse(source);

        let root = tree.root_node();

        // Walk the tree
        let mut cursor = root.walk();

        // Should be able to navigate to children
        assert!(cursor.goto_first_child(), "Should have first child");

        // Should be able to navigate back
        assert!(cursor.goto_parent(), "Should return to parent");
    }

    #[test]
    fn test_empty_source() {
        let source = "";
        let tree = FshGritTree::parse(source);

        let root = tree.root_node();
        assert_eq!(
            root.children().count(),
            0,
            "Empty source should have no children"
        );
    }

    #[test]
    fn test_trivia_preservation() {
        let source = r#"
// This is a comment
Profile: MyPatient  // Inline comment
  Parent: Patient
"#;
        let tree = FshGritTree::parse(source);

        // The CST should preserve all trivia
        let root = tree.root_node();
        let reconstructed = root.text().unwrap();
        assert_eq!(
            reconstructed.as_ref(),
            source,
            "Should preserve comments and whitespace"
        );
    }
}
