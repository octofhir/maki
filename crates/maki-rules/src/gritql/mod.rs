//! GritQL integration for FSH linting - CST-based implementation
//!
//! This module provides full GritQL pattern matching support for FSH files
//! using our Rowan-based CST (Concrete Syntax Tree).
//!
//! # Architecture
//!
//! - **CST Adapter** (`cst_adapter`): Wraps Rowan nodes to implement GritQL's AstNode trait
//! - **CST Tree** (`cst_tree`): Implements GritQL's Ast trait over our CST
//! - **Language** (`cst_language`): Tells GritQL how to work with FSH syntax
//! - **Executor**: Compiles and executes GritQL patterns against FSH code
//! - **Built-ins**: FSH-specific functions for use in GritQL patterns
//!
//! # Example
//!
//! ```ignore
//! use maki_rules::gritql::{FshGritTree, GritQLCompiler};
//!
//! // Parse FSH to queryable tree
//! let tree = FshGritTree::parse("Profile: MyPatient\nParent: Patient");
//!
//! // Compile a pattern
//! let compiler = GritQLCompiler::new()?;
//! let pattern = compiler.compile_pattern(
//!     "Profile: $name where $name == `MyPatient`",
//!     "test-rule"
//! )?;
//!
//! // Execute and get matches
//! let matches = pattern.execute("Profile: MyPatient", "test.fsh")?;
//! ```

// CST-based implementation
pub mod cst_adapter;
pub mod cst_language;
pub mod cst_tree;

// QueryContext - bridges our CST with grit-pattern-matcher
pub mod query_context;

// GritQL parser and compiler
pub mod compiler;
pub mod parser;

// Executor and built-ins
pub mod builtins;
pub mod executor;

// Rule loading and registry
pub mod loader;
pub mod registry;

// Re-export main types
pub use builtins::register_fsh_builtins;
pub use cst_adapter::{FshGritCursor, FshGritNode};
pub use cst_language::FshTargetLanguage;
pub use cst_tree::FshGritTree;
pub use executor::{CompiledGritQLPattern, GritQLCompiler, GritQLMatch, MatchRange};
pub use loader::GritQLRuleLoader;
pub use registry::GritQLRegistry;
