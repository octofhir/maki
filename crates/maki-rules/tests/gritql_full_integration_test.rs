//! Full GritQL Integration Tests with Snapshot Testing
//!
//! Tests the complete GritQL pipeline:
//! - Parser: GritQL syntax → AST
//! - Compiler: AST → Pattern<FshQueryContext>
//! - QueryContext: All trait implementations
//! - Pattern matching against FSH CST

use grit_pattern_matcher::binding::Binding;
use grit_pattern_matcher::pattern::ResolvedPattern;
use grit_util::{Ast, AstNode};
use insta::assert_yaml_snapshot;
use maki_core::cst::parse_fsh;
use maki_rules::gritql::{
    compiler::PatternCompiler,
    cst_tree::FshGritTree,
    parser::GritQLParser,
    query_context::{FshBinding, FshQueryContext, FshResolvedPattern},
};

#[test]
fn test_parser_simple_node_pattern() {
    let mut parser = GritQLParser::new("profile");
    let result = parser.parse();

    assert!(result.is_ok());
    assert_yaml_snapshot!("parser_simple_node", result.unwrap());
}

#[test]
fn test_parser_where_clause() {
    let mut parser = GritQLParser::new("profile where { description }");
    let result = parser.parse();

    assert!(result.is_ok());
    assert_yaml_snapshot!("parser_where_clause", result.unwrap());
}

#[test]
fn test_parser_variable() {
    let mut parser = GritQLParser::new("$name");
    let result = parser.parse();

    assert!(result.is_ok());
    assert_yaml_snapshot!("parser_variable", result.unwrap());
}

#[test]
fn test_parser_assignment() {
    let mut parser = GritQLParser::new("$name = profile");
    let result = parser.parse();

    assert!(result.is_ok());
    assert_yaml_snapshot!("parser_assignment", result.unwrap());
}

#[test]
fn test_parser_not_pattern() {
    let mut parser = GritQLParser::new("not profile");
    let result = parser.parse();

    assert!(result.is_ok());
    assert_yaml_snapshot!("parser_not", result.unwrap());
}

#[test]
fn test_compiler_simple_node() {
    let mut parser = GritQLParser::new("profile");
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
    // Pattern compilation successful
}

#[test]
fn test_compiler_where_clause() {
    let mut parser = GritQLParser::new("profile where { description }");
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
    // Where clause compilation successful
}

#[test]
fn test_compiler_variable() {
    let mut parser = GritQLParser::new("$name");
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
    // Variable compilation successful
}

#[test]
fn test_compiler_assignment() {
    let mut parser = GritQLParser::new("$name = profile");
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
    // Assignment compilation successful
}

#[test]
fn test_cst_integration_parse_simple_profile() {
    let source = "Profile: MyPatient\nParent: Patient\nTitle: \"My Patient Profile\"";
    let (cst, errors) = parse_fsh(source);

    assert!(errors.is_empty(), "Parse errors: {errors:?}");

    let tree = FshGritTree::new(cst, source.to_string());
    let root = tree.root_node();

    assert!(root.text().is_ok());
}

#[test]
fn test_cst_tree_creation() {
    let source = "Profile: TestProfile\nParent: Patient";
    let (cst, errors) = parse_fsh(source);

    assert!(errors.is_empty());

    let tree = FshGritTree::new(cst, source.to_string());

    assert_eq!(tree.source().as_ref(), source);
}

#[test]
fn test_resolved_pattern_from_node_binding() {
    let source = "Profile: TestProfile";
    let (cst, _) = parse_fsh(source);
    let tree = FshGritTree::new(cst, source.to_string());
    let node = tree.root_node();

    let resolved = FshResolvedPattern::from_node_binding(node);

    assert!(resolved.is_binding());
}

#[test]
fn test_resolved_pattern_from_string() {
    let resolved = FshResolvedPattern::from_string("test".to_string());

    assert!(!resolved.is_binding());
    assert!(!resolved.is_list());
}

#[test]
fn test_resolved_pattern_from_constant() {
    use grit_pattern_matcher::constant::Constant;

    let resolved = FshResolvedPattern::from_constant(Constant::Integer(42));

    assert!(!resolved.is_binding());
}

#[test]
fn test_binding_from_node() {
    let source = "Profile: TestProfile";
    let (cst, _) = parse_fsh(source);
    let tree = FshGritTree::new(cst, source.to_string());
    let node = tree.root_node();

    let binding = FshBinding::Node(node.clone());

    assert!(binding.singleton().is_some());
}

#[test]
fn test_binding_from_constant() {
    use grit_pattern_matcher::constant::Constant;

    static CONST: Constant = Constant::Integer(123);
    let binding = FshBinding::Constant(&CONST);

    assert!(binding.as_constant().is_some());
}

#[test]
fn test_full_pipeline_parse_compile() {
    // Test the full pipeline: Parse GritQL -> Compile -> Pattern ready for execution
    let gritql = "profile where { description }";

    let mut parser = GritQLParser::new(gritql);
    let parsed = parser.parse().expect("Parse failed");

    let mut compiler = PatternCompiler::new();
    let _compiled = compiler.compile(&parsed).expect("Compile failed");

    // If we get here, the pattern is ready for execution
    assert_yaml_snapshot!("full_pipeline_parse_compile", format!("{:?}", parsed));
}

#[test]
fn test_syntax_kind_mapping() {
    let mut compiler = PatternCompiler::new();

    // Test various FSH syntax kinds
    let test_cases = vec![
        ("profile", true),
        ("extension", true),
        ("valueset", true),
        ("codesystem", true),
        ("instance", true),
        ("unknown_kind", false),
    ];

    for (kind, should_succeed) in test_cases {
        let mut parser = GritQLParser::new(kind);
        let pattern = parser.parse().unwrap();
        let result = compiler.compile(&pattern);

        assert_eq!(result.is_ok(), should_succeed, "Failed for kind: {kind}");
    }
}

#[test]
fn test_complex_pattern_and() {
    let gritql = "profile"; // Would need AND support in parser

    let mut parser = GritQLParser::new(gritql);
    let result = parser.parse();

    assert!(result.is_ok());
}

#[test]
fn test_complex_pattern_or() {
    let gritql = "profile"; // Would need OR support in parser

    let mut parser = GritQLParser::new(gritql);
    let result = parser.parse();

    assert!(result.is_ok());
}

#[test]
fn test_nested_where_clauses() {
    let gritql = "profile where { description }";

    let mut parser = GritQLParser::new(gritql);
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
}

#[test]
fn test_variable_scoping() {
    let gritql = "$x = profile";

    let mut parser = GritQLParser::new(gritql);
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
}

#[test]
fn test_multiple_variables() {
    // Test that variables get unique indices
    let gritql1 = "$x";
    let gritql2 = "$y";

    let mut compiler = PatternCompiler::new();

    let mut parser1 = GritQLParser::new(gritql1);
    let pattern1 = parser1.parse().unwrap();
    compiler.compile(&pattern1).unwrap();

    let mut parser2 = GritQLParser::new(gritql2);
    let pattern2 = parser2.parse().unwrap();
    compiler.compile(&pattern2).unwrap();

    // Both variables should have been registered
}

#[test]
fn test_predicate_field_exists() {
    let gritql = "profile where { description }";

    let mut parser = GritQLParser::new(gritql);
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_ok());
    assert_yaml_snapshot!("predicate_field_exists", format!("{:?}", pattern));
}

#[test]
fn test_predicate_not() {
    let gritql = "profile where { description }"; // Parser would need NOT predicate support

    let mut parser = GritQLParser::new(gritql);
    let result = parser.parse();

    assert!(result.is_ok());
}

#[test]
fn test_error_invalid_syntax_kind() {
    let gritql = "invalid_syntax_kind_that_does_not_exist";

    let mut parser = GritQLParser::new(gritql);
    let pattern = parser.parse().unwrap();

    let mut compiler = PatternCompiler::new();
    let result = compiler.compile(&pattern);

    assert!(result.is_err());
}

#[test]
fn test_resolved_pattern_truthy() {
    use grit_pattern_matcher::constant::Constant;
    use grit_pattern_matcher::pattern::{FileRegistry, ResolvedPattern, State};
    use maki_rules::gritql::cst_language::FshTargetLanguage;

    let resolved = FshResolvedPattern::from_constant(Constant::Boolean(true));

    let lang = FshTargetLanguage;
    let registry = FileRegistry::new_from_paths(vec![]);
    let mut state: State<FshQueryContext> = State::new(vec![], registry);

    let is_truthy = resolved.is_truthy(&mut state, &lang).unwrap();
    assert!(is_truthy);
}

#[test]
fn test_resolved_pattern_text() {
    use grit_pattern_matcher::constant::Constant;
    use grit_pattern_matcher::pattern::{FileRegistry, ResolvedPattern};
    use maki_rules::gritql::cst_language::FshTargetLanguage;

    let resolved = FshResolvedPattern::from_constant(Constant::String("test".to_string()));

    let lang = FshTargetLanguage;
    let files = FileRegistry::new_from_paths(vec![]);

    let text = resolved.text(&files, &lang).unwrap();
    assert_eq!(text.as_ref(), "test");
}

#[test]
fn test_cst_adapter_node_methods() {
    let source = "Profile: TestProfile\nParent: Patient";
    let (cst, _) = parse_fsh(source);
    let tree = FshGritTree::new(cst, source.to_string());
    let node = tree.root_node();

    // Test various node methods
    assert!(node.byte_range().start < node.byte_range().end);
    assert!(!node.text_content().is_empty());
}

#[test]
fn test_comprehensive_fsh_parse_and_tree_creation() {
    let fsh_samples = [
        "Profile: SimpleProfile\nParent: Patient",
        "Extension: MyExtension\nId: my-ext",
        "ValueSet: MyValueSet\nTitle: \"My VS\"",
        "CodeSystem: MyCS",
        "Instance: MyInstance\nInstanceOf: Patient",
    ];

    for (idx, source) in fsh_samples.iter().enumerate() {
        let (cst, errors) = parse_fsh(source);
        assert!(
            errors.is_empty(),
            "Parse errors for sample {idx}: {errors:?}"
        );

        let tree = FshGritTree::new(cst, source.to_string());
        let root = tree.root_node();

        assert!(
            !root.text_content().is_empty(),
            "Empty text for sample {idx}"
        );
        assert_eq!(tree.source().as_ref(), *source);
    }
}
