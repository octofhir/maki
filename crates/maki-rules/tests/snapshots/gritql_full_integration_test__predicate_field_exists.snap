---
source: crates/maki-rules/tests/gritql_full_integration_test.rs
expression: "format!(\"{:?}\", pattern)"
---
"Where(NodeKind(\"profile\"), FieldExists(\"description\"))"
